# feature scaling example code

In [1]:
import pandas as pd
import numpy as np

from sklearn import preprocessing

In [2]:
NBA_FILE_PATH = 'C:/Users/kth25/ai_dataset/NBA_player_of_the_week.csv'
nba_player_of_the_week_df = pd.read_csv(NBA_FILE_PATH)

In [3]:
nba_player_of_the_week_df.head()

Player                    Team Conference          Date Position  \
0    Jayson Tatum          Boston Celtics       East  Feb 10, 2020       SF   
1    Nikola Jokic          Denver Nuggets       West  Feb 10, 2020        C   
2    Jaylen Brown          Boston Celtics       East   Feb 3, 2020       SF   
3  Damian Lillard  Portland Trail Blazers       West   Feb 3, 2020        G   
4   Pascal Siakam         Toronto Raptors       East  Jan 27, 2020        F   

  Height  Weight  Age  Draft Year  Seasons in league     Season  Season short  \
0    6'8     208   21        2017                  2  2019-2020          2020   
1    7'0     250   25        2014                  4  2019-2020          2020   
2    6'7     220   23        2016                  3  2019-2020          2020   
3    6'3     195   29        2012                  7  2019-2020          2020   
4    6'9     230   25        2016                  3  2019-2020          2020   

           Pre-draft Team  Real_value  Height CM  Weight KG  Last Season  
0                    Duke         0.5        203         94            1  
1  KK Mega Bemax (Serbia)         0.5        213        113            1  
2              California         0.5        201         99            1  
3             Weber State         0.5        190         88            1  
4        New Mexico State         0.5        206        104            1

In [4]:
nba_player_of_the_week_df.describe()

Weight          Age   Draft Year  Seasons in league  Season short  \
count  1340.000000  1340.000000  1340.000000        1340.000000   1340.000000   
mean    224.567164    26.738060  1996.287313           5.740299   2003.156716   
std      30.798885     3.400683    11.253558           3.293421     11.470164   
min     150.000000    19.000000  1965.000000           0.000000   1980.000000   
25%     205.000000    24.000000  1987.000000           3.000000   1994.000000   
50%     220.000000    26.000000  1998.000000           5.000000   2005.000000   
75%     250.000000    29.000000  2005.000000           8.000000   2013.000000   
max     325.000000    40.000000  2018.000000          17.000000   2020.000000   

        Real_value    Height CM    Weight KG  Last Season  
count  1340.000000  1340.000000  1340.000000  1340.000000  
mean      0.686940   201.071642   101.384328     0.023881  
std       0.242007     9.367970    14.011226     0.152734  
min       0.500000   175.000000    68.000000     0.000000  
25%       0.500000   193.000000    93.000000     0.000000  
50%       0.500000   201.000000    99.000000     0.000000  
75%       1.000000   208.000000   113.000000     0.000000  
max       1.000000   229.000000   147.000000     1.000000

In [5]:
height_weight_age_df = nba_player_of_the_week_df[['Height CM','Weight KG','Age']]
height_weight_age_df.head()

Height CM  Weight KG  Age
0        203         94   21
1        213        113   25
2        201         99   23
3        190         88   29
4        206        104   25

### MinMaxScaler를 이용한 Normalization

MinMaxNormalization이란 data의 최솟값과 최댓값을 이용하여 data를 0~1사이의 범위로 scaling 해주는 방법을 말한다!

x_new = (x_old - x_min) / (x_max - x_min)

In [6]:
scaler = preprocessing.MinMaxScaler()
normalized_data = scaler.fit_transform(height_weight_age_df)
normalized_data

array([[0.51851852, 0.32911392, 0.0952381 ],
       [0.7037037 , 0.56962025, 0.28571429],
       [0.48148148, 0.39240506, 0.19047619],
       ...,
       [0.48148148, 0.37974684, 0.23809524],
       [0.38888889, 0.21518987, 0.23809524],
       [0.42592593, 0.27848101, 0.52380952]])

In [7]:
normalized_df = pd.DataFrame(normalized_data, columns=['Height','Weight','Age'])

In [8]:
normalized_df.describe()

Height       Weight          Age
count  1340.000000  1340.000000  1340.000000
mean      0.482808     0.422586     0.368479
std       0.173481     0.177357     0.161937
min       0.000000     0.000000     0.000000
25%       0.333333     0.316456     0.238095
50%       0.481481     0.392405     0.333333
75%       0.611111     0.569620     0.476190
max       1.000000     1.000000     1.000000

### StandardScaler를 이용한 Normalization

Standardization은 data를 평균이 0, 표준 편차가 1이 되도록 표준화 시켜주는 방법을 말한다!

x_new = (x_old - x_mean) / (표준편차)

In [9]:
scaler = preprocessing.StandardScaler()
standardized_data = scaler.fit_transform(height_weight_age_df)
standardized_data

array([[ 0.20592274, -0.52722617, -1.68795564],
       [ 1.27378837,  0.82933556, -0.51128218],
       [-0.00765038, -0.17023624, -1.09961891],
       ...,
       [-0.00765038, -0.24163423, -0.80545055],
       [-0.54158319, -1.16980804, -0.80545055],
       [-0.32801007, -0.81281811,  0.95955965]])

In [10]:
standardized_df = pd.DataFrame(standardized_data, columns=['Height', 'Weight', 'Age'])

In [11]:
standardized_df.describe()

Height        Weight           Age
count  1.340000e+03  1.340000e+03  1.340000e+03
mean  -1.070288e-15 -7.049088e-16 -3.557685e-16
std    1.000373e+00  1.000373e+00  1.000373e+00
min   -2.784101e+00 -2.383574e+00 -2.276292e+00
25%   -8.619429e-01 -5.986242e-01 -8.054505e-01
50%   -7.650381e-03 -1.702362e-01 -2.171138e-01
75%    7.398556e-01  8.293356e-01  6.653913e-01
max    2.982373e+00  3.256867e+00  3.901243e+00

# One-hot Encoding Example

In [12]:
TITANIC_FILE_PATH = 'C:/Users/kth25/ai_dataset/titanic.csv'
titanic_df = pd.read_csv(TITANIC_FILE_PATH)
titanic_df.head()

Unnamed: 0  Survived  Pclass  \
0           0         0       3   
1           1         1       1   
2           2         1       3   
3           3         1       1   
4           4         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [13]:
titanic_sex_embarked = titanic_df[['Sex','Embarked']]

In [14]:
titanic_sex_embarked.head()

Sex Embarked
0    male        S
1  female        C
2  female        S
3  female        S
4    male        S

In [15]:
one_hot_encoded_df = pd.get_dummies(titanic_sex_embarked)

In [16]:
one_hot_encoded_df.head()

Sex_female  Sex_male  Embarked_C  Embarked_Q  Embarked_S
0           0         1           0           0           1
1           1         0           1           0           0
2           1         0           0           0           1
3           1         0           0           0           1
4           0         1           0           0           1

In [17]:
# 원하는 열들만 one-hot Encoding을 하고 싶은 경우
one_hot_encoded_df = pd.get_dummies(data=titanic_df,columns=['Sex','Embarked'])
one_hot_encoded_df.head()

Unnamed: 0  Survived  Pclass  \
0           0         0       3   
1           1         1       1   
2           2         1       3   
3           3         1       1   
4           4         0       3   

                                                Name   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0      1      0   
2                             Heikkinen, Miss. Laina  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0      1      0   
4                           Allen, Mr. William Henry  35.0      0      0   

             Ticket     Fare Cabin  Sex_female  Sex_male  Embarked_C  \
0         A/5 21171   7.2500   NaN           0         1           0   
1          PC 17599  71.2833   C85           1         0           1   
2  STON/O2. 3101282   7.9250   NaN           1         0           0   
3            113803  53.1000  C123           1         0           0   
4            373450   8.0500   NaN           0         1           0   

   Embarked_Q  Embarked_S  
0           0           1  
1           0           0  
2           0           1  
3           0           1  
4           0           1

# Overfitting Example

In [18]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import PolynomialFeatures

from math import sqrt

In [20]:
ADMISSION_FILE_PATH = 'C:/Users/kth25/ai_dataset/admission_data.csv'
admission_df = pd.read_csv(ADMISSION_FILE_PATH).drop('Serial No.',axis=1)

In [21]:
admission_df.head()

GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  Research  \
0        337          118                  4  4.5   4.5  9.65         1   
1        324          107                  4  4.0   4.5  8.87         1   
2        316          104                  3  3.0   3.5  8.00         1   
3        322          110                  3  3.5   2.5  8.67         1   
4        314          103                  2  2.0   3.0  8.21         0   

   Chance of Admit   
0              0.92  
1              0.76  
2              0.72  
3              0.80  
4              0.65

In [24]:
X = admission_df.drop(['Chance of Admit '],axis=1)

polynomial_transformer = PolynomialFeatures(6)
polynomial_features = polynomial_transformer.fit_transform(X.values)
features = polynomial_transformer.get_feature_names(X.columns)

X = pd.DataFrame(polynomial_features, columns=features)
X.head()

1  GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  Research  \
0  1.0      337.0        118.0                4.0  4.5   4.5  9.65       1.0   
1  1.0      324.0        107.0                4.0  4.0   4.5  8.87       1.0   
2  1.0      316.0        104.0                3.0  3.0   3.5  8.00       1.0   
3  1.0      322.0        110.0                3.0  3.5   2.5  8.67       1.0   
4  1.0      314.0        103.0                2.0  2.0   3.0  8.21       0.0   

   GRE Score^2  GRE Score TOEFL Score  ...  LOR  CGPA^2 Research^3  \
0     113569.0                39766.0  ...               419.05125   
1     104976.0                34668.0  ...               354.04605   
2      99856.0                32864.0  ...               224.00000   
3     103684.0                35420.0  ...               187.92225   
4      98596.0                32342.0  ...                 0.00000   

   LOR  CGPA Research^4  LOR  Research^5         CGPA^6  CGPA^5 Research  \
0                43.425              4.5  807539.696082     83682.870060   
1                39.915              4.5  487014.306256     54905.784245   
2                28.000              3.5  262144.000000     32768.000000   
3                21.675              2.5  424731.610940     48988.651781   
4                 0.000              0.0  306237.903347         0.000000   

   CGPA^4 Research^2  CGPA^3 Research^3  CGPA^2 Research^4  CGPA Research^5  \
0        8671.800006         898.632125            93.1225             9.65   
1        6190.054594         697.864103            78.6769             8.87   
2        4096.000000         512.000000            64.0000             8.00   
3        5650.363527         651.714363            75.1689             8.67   
4           0.000000           0.000000             0.0000             0.00   

   Research^6  
0         1.0  
1         1.0  
2         1.0  
3         1.0  
4         0.0  

[5 rows x 1716 columns]

In [25]:
y = admission_df[['Chance of Admit ']]
y.head()

Chance of Admit 
0              0.92
1              0.76
2              0.72
3              0.80
4              0.65

In [26]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=5)

In [27]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [28]:
y_train_predict = model.predict(X_train)
y_test_predict = model.predict(X_test)

In [30]:
mse = mean_squared_error(y_train, y_train_predict)

print('training set 에서의 성능')
print('==========================')
print(sqrt(mse))
print()

mse = mean_squared_error(y_test, y_test_predict)
print('test set 에서의 성능')
print('==========================')
print(sqrt(mse))

training set 에서의 성능
0.0015048194399234639

test set 에서의 성능
5.090721961126577


# overfitting 해결하기 _ by L1 정규화

In [32]:
from sklearn.linear_model import Lasso
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import PolynomialFeatures

from math import sqrt

In [20]:
ADMISSION_FILE_PATH = 'C:/Users/kth25/ai_dataset/admission_data.csv'
admission_df = pd.read_csv(ADMISSION_FILE_PATH).drop('Serial No.',axis=1)

In [21]:
admission_df.head()

GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  Research  \
0        337          118                  4  4.5   4.5  9.65         1   
1        324          107                  4  4.0   4.5  8.87         1   
2        316          104                  3  3.0   3.5  8.00         1   
3        322          110                  3  3.5   2.5  8.67         1   
4        314          103                  2  2.0   3.0  8.21         0   

   Chance of Admit   
0              0.92  
1              0.76  
2              0.72  
3              0.80  
4              0.65

In [24]:
X = admission_df.drop(['Chance of Admit '],axis=1)

polynomial_transformer = PolynomialFeatures(6)
polynomial_features = polynomial_transformer.fit_transform(X.values)
features = polynomial_transformer.get_feature_names(X.columns)

X = pd.DataFrame(polynomial_features, columns=features)
X.head()

1  GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  Research  \
0  1.0      337.0        118.0                4.0  4.5   4.5  9.65       1.0   
1  1.0      324.0        107.0                4.0  4.0   4.5  8.87       1.0   
2  1.0      316.0        104.0                3.0  3.0   3.5  8.00       1.0   
3  1.0      322.0        110.0                3.0  3.5   2.5  8.67       1.0   
4  1.0      314.0        103.0                2.0  2.0   3.0  8.21       0.0   

   GRE Score^2  GRE Score TOEFL Score  ...  LOR  CGPA^2 Research^3  \
0     113569.0                39766.0  ...               419.05125   
1     104976.0                34668.0  ...               354.04605   
2      99856.0                32864.0  ...               224.00000   
3     103684.0                35420.0  ...               187.92225   
4      98596.0                32342.0  ...                 0.00000   

   LOR  CGPA Research^4  LOR  Research^5         CGPA^6  CGPA^5 Research  \
0                43.425              4.5  807539.696082     83682.870060   
1                39.915              4.5  487014.306256     54905.784245   
2                28.000              3.5  262144.000000     32768.000000   
3                21.675              2.5  424731.610940     48988.651781   
4                 0.000              0.0  306237.903347         0.000000   

   CGPA^4 Research^2  CGPA^3 Research^3  CGPA^2 Research^4  CGPA Research^5  \
0        8671.800006         898.632125            93.1225             9.65   
1        6190.054594         697.864103            78.6769             8.87   
2        4096.000000         512.000000            64.0000             8.00   
3        5650.363527         651.714363            75.1689             8.67   
4           0.000000           0.000000             0.0000             0.00   

   Research^6  
0         1.0  
1         1.0  
2         1.0  
3         1.0  
4         0.0  

[5 rows x 1716 columns]

In [25]:
y = admission_df[['Chance of Admit ']]
y.head()

Chance of Admit 
0              0.92
1              0.76
2              0.72
3              0.80
4              0.65

In [26]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=5)

In [33]:
# max_iter은 경사하강법을 최대한 얼만큼 할지
# L2 정규화를 쓰고 싶으면 Lasso를 Ridge로만 바꿔주면 돼!!

model = Lasso(alpha = 0.001, max_iter = 1000, normalize=True)
model.fit(X_train, y_train)

Lasso(alpha=0.001, normalize=True)

In [34]:
y_train_predict = model.predict(X_train)
y_test_predict = model.predict(X_test)

In [35]:
mse = mean_squared_error(y_train, y_train_predict)

print('training set 에서의 성능')
print('==========================')
print(sqrt(mse))
print()

mse = mean_squared_error(y_test, y_test_predict)
print('test set 에서의 성능')
print('==========================')
print(sqrt(mse))

training set 에서의 성능
0.06336620966147143

test set 에서의 성능
0.060077190926892544


### 결과 정리!!

1. overfitting이 일어났을 때 : training set에 대한 성능은 월등히 뛰어나지만, test set에서의 성능은 현저히 낮아지는 것을 확인할 수 있다.

training set 에서의 성능
==========================
*0.0015048194399234639*

test set 에서의 성능
==========================
*5.090721961126577*

2. Regularization을 적용했을 때 : training set에 대한 성능은 조금 줄어들었지만, test set에서의 성능은 월등히 높아진 것을 확인할 수 있다.

training set 에서의 성능
==========================
*0.06336620966147143*

test set 에서의 성능
==========================
*0.060077190926892544*